pyiron/atomistics/structure/atoms.py

```diff
@@ -1744,120 +1688,52 @@ def _get_neighbors(
-                ic += 1
-            neighbor_obj.indices = ind_lst
-            neighbor_obj.distances = d_lst
-            neighbor_obj.vecs = v_lst
-            return neighbor_obj
-
-        # include periodic boundaries
-        # translate radius in boundary layer with relative coordinates
-        # TODO: introduce more rigoros definition
-        radius = 3 * num_neighbors ** (1.0 / 3.0)
-        rel_width = [radius / np.sqrt(np.dot(a_i, a_i)) for a_i in self.cell]
-        rel_width_scalar = np.max(rel_width)
-
-        # construct cell with additional atoms bounding original cell
-        boundary_atoms = self.get_boundary_region(rel_width_scalar)
-        extended_cell = self + boundary_atoms
-
-        # build index to map boundary atoms back to original cell
-        map_to_cell = np.append(np.arange(len(self)), self._ia_bounds)
-
-        # transfer relative to absolute coordinates
-        tree = cKDTree(extended_cell.positions)
-        if id_list is None:
-            positions = self.positions
+        if all(self.pbc==False):
+            width = 0
+        elif cutoff_radius!=np.inf:
+            width = cutoff_radius
         else:
-            positions = np.array([self.positions[i] for i in id_list])
-        # print ("len positions: ", len(positions))
-        if cutoff_radius is None:
-            neighbors = tree.query(positions, k=num_neighbors)
+            prefactor = [1, 1/np.pi, 4/(3*np.pi)]
+            prefactor = prefactor[sum(self.pbc)-1]
+            width = np.prod((np.linalg.norm(self.cell, axis=-1)-np.ones(3))*self.pbc+np.ones(3))
+            width *= prefactor*np.max([num_neighbors, 8])/len(self)
+            width = boundary_width_factor*width**(1/np.sum(self.pbc))
+        neighbor_obj._boundary_layer_width = width
+        if (width<0.5*np.min(self.cell.diagonal())
+            and np.isclose(np.linalg.norm(self.cell-np.eye(3)*self.cell.diagonal()), 0)
+            and np.all(self.pbc)
+            and cutoff_radius==np.inf):
+            boxsize = self.cell.diagonal()
+            extended_positions, indices = self.get_extended_positions(0)
+            extended_positions /= self.cell.diagonal()
+            extended_positions -= np.floor(extended_positions)
+            extended_positions *= self.cell.diagonal()
         else:
-            neighbors = tree.query(
-                positions, k=num_neighbors, distance_upper_bound=cutoff_radius
-            )
-
-        # print ("neighbors: ", neighbors)
-
+            extended_positions, indices = self.get_extended_positions(width)
+        if len(extended_positions) < num_neighbors and cutoff_radius==np.inf:
+            raise ValueError('num_neighbors too large - make boundary_width_factor larger and/or make num_neighbors smaller')
+        tree = cKDTree(extended_positions, boxsize=boxsize)
+        if id_list is None:
+            id_list = np.arange(len(self.positions))
+        positions = self.positions[np.array(id_list)]
```

Comment:
Why not
```python
if id_list is None:
    positions = self.positions
else:
    positions = self.positions[np.asarray(id_list)]
```
That avoids an array allocation in the common case that `id_list == None`.

Reply:
`id_list` is used again later in the same function

Reply:
I was hoping to avoid that usage by the other suggestion I made there, but didn't see it was used again in the branch when `t_vec` is True, my bad.